that last change broke the unit tests :see_no_evil: 

```
--- FAIL: Test (0.02s)
    --- FAIL: Test/EgressGatewayTestSuite (0.02s)
        --- FAIL: Test/EgressGatewayTestSuite/TestEgressGatewayManager (0.02s)
            manager_privileged_test.go:630: 
                ... value *fmt.wrapError = &fmt.wrapError{msg:"lookup: key does not exist", err:(*sys.syscallError)(0x3b53c20)} ("lookup: key does not exist")
```

looking into

--- a/pkg/egressgateway/endpoint.go
+++ b/pkg/egressgateway/endpoint.go
@@ -38,6 +38,10 @@ func getEndpointMetadata(endpoint *k8sTypes.CiliumEndpoint, identityLabels label
 		return nil, fmt.Errorf("endpoint has no networking metadata")
 	}
 
+	if len(endpoint.Networking.Addressing) == 0 {
+		return nil, fmt.Errorf("failed to get valid endpoint IPs")
+	}
+
 	for _, pair := range endpoint.Networking.Addressing {
 		if pair.IPV4 != "" {
 			ipv4s = append(ipv4s, net.ParseIP(pair.IPV4).To4())

--- a/pkg/egressgateway/manager.go
+++ b/pkg/egressgateway/manager.go
@@ -345,13 +345,6 @@ func (manager *Manager) addEndpoint(id types.NamespacedName) {
 	// encounter a failure it cannot be retried
 	delete(manager.pendingEndpointEvents, id)
 
-	if len(endpoint.Networking.Addressing) == 0 {
-		logger.WithError(err).
-			Error("Failed to get valid endpoint IPs, skipping update to egress policy.")
-
-		return
-	}
-
 	if epData, err = getEndpointMetadata(endpoint, identityLabels); err != nil {
 		logger.WithError(err).
 			Error("Failed to get valid endpoint metadata, skipping update to egress policy.")

--- a/pkg/egressgateway/manager.go
+++ b/pkg/egressgateway/manager.go
@@ -8,11 +8,14 @@ import (
 	"fmt"
 	"net"
 	"sort"
+	"sync"
+	"time"
 
 	"github.com/sirupsen/logrus"
 	"github.com/spf13/pflag"
 	"github.com/vishvananda/netlink"
 	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
 
 	"github.com/cilium/cilium/pkg/datapath/linux/probes"
 	"github.com/cilium/cilium/pkg/hive"
@@ -101,6 +104,17 @@ type Manager struct {
 	// epDataStore stores endpointId to endpoint metadata mapping
 	epDataStore map[endpointID]*endpointMetadata
 
+	// pendingEndpointEvents stores the k8s CiliumEndpoint add/update events
+	// which still need to be processed by the manager, either because we
+	// just received the event, or because the processing failed due to the
+	// manager being unable to resolve the endpoint identity to a set of
+	// labels
+	pendingEndpointEvents map[endpointID]*k8sTypes.CiliumEndpoint
+
+	// endpointEventsQueue is a workqueue of CiliumEndpoint IDs that need to
+	// be processed by the manager
+	endpointEventsQueue workqueue.RateLimitingInterface
+
 	// identityAllocator is used to fetch identity labels for endpoint updates
 	identityAllocator identityCache.IdentityAllocator
 
@@ -130,16 +144,27 @@ func NewEgressGatewayManager(p Params) *Manager {
 		return nil
 	}
 
+	// here we try to mimic the same exponential backoff retry logic used by
+	// the identity allocator, where the minimum retry timeout is set to 20
+	// milliseconds and the max number of attempts is 16 (so 20ms * 2^16 ==
+	// ~20 minutes)
+	rateLimiter := workqueue.NewItemExponentialFailureRateLimiter(time.Millisecond*20, time.Minute*20)
+	endpointEventRetryQueue := workqueue.NewRateLimitingQueueWithConfig(rateLimiter, workqueue.RateLimitingQueueConfig{})
+
 	manager := &Manager{
 		cacheStatus:             p.CacheStatus,
 		nodeDataStore:           make(map[string]nodeTypes.Node),
 		policyConfigs:           make(map[policyID]*PolicyConfig),
 		policyConfigsBySourceIP: make(map[string][]*PolicyConfig),
 		epDataStore:             make(map[endpointID]*endpointMetadata),
+		pendingEndpointEvents:   make(map[endpointID]*k8sTypes.CiliumEndpoint),
+		endpointEventsQueue:     endpointEventRetryQueue,
 		identityAllocator:       p.IdentityAllocator,
 		installRoutes:           p.Config.InstallEgressGatewayRoutes,
 		policyMap:               p.PolicyMap,
 	}
+
+	var wg sync.WaitGroup
 
 	ctx, cancel := context.WithCancel(context.Background())
 	p.Lifecycle.Append(hive.Hook{
@@ -149,10 +174,13 @@ func NewEgressGatewayManager(p Params) *Manager {
 			}
 
 			manager.runReconciliationAfterK8sSync(ctx)
+			manager.processCiliumEndpoints(ctx, &wg)
 			return nil
 		},
 		OnStop: func(hc hive.HookContext) error {
 			cancel()
+
+			wg.Wait()
 			return nil
 		},
 	})
@@ -186,6 +214,51 @@ func (manager *Manager) runReconciliationAfterK8sSync(ctx context.Context) {
 			manager.reconcile(eventK8sSyncDone)
 			manager.Unlock()
 		case <-ctx.Done():
+		}
+	}()
+}
+
+// processCiliumEndpoints spawns a goroutine that:
+//   - consumes the endpoint IDs returned by the endpointEventsQueue workqueue
+//   - processes the CiliumEndpoints stored in pendingEndpointEvents for these
+//     endpoint IDs
+//   - in case the endpoint ID -> labels resolution fails, it adds back the
+//     event to the workqueue so that it can be retried with an exponential
+//     backoff
+func (manager *Manager) processCiliumEndpoints(ctx context.Context, wg *sync.WaitGroup) {
+	wg.Add(1)
+
+	go func() {
+		defer wg.Done()
+
+		retryQueue := manager.endpointEventsQueue
+		go func() {
+			select {
+			case <-ctx.Done():
+				retryQueue.ShutDown()
+			}
+		}()
+
+		for {
+			item, shutdown := retryQueue.Get()
+			if shutdown {
+				break
+			}
+			endpointID := item.(types.NamespacedName)
+
+			manager.addEndpoint(endpointID)
+
+			if manager.endpointEventIsPending(endpointID) {
+				// if the endpoint event is still pending it means the manager
+				// failed to resolve the endpoint ID to a set of labels, so add back
+				// the item to the queue
+				manager.endpointEventsQueue.AddRateLimited(endpointID)
+			} else {
+				// otherwise just remove it
+				manager.endpointEventsQueue.Forget(endpointID)
+			}
+
+			manager.endpointEventsQueue.Done(endpointID)
 		}
 	}()
 }
@@ -233,8 +306,15 @@ func (manager *Manager) OnDeleteEgressPolicy(configID policyID) {
 	manager.reconcile(eventDeletePolicy)
 }
 
-// OnUpdateEndpoint is the event handler for endpoint additions and updates.
-func (manager *Manager) OnUpdateEndpoint(endpoint *k8sTypes.CiliumEndpoint) {
+func (manager *Manager) endpointEventIsPending(id types.NamespacedName) bool {
+	manager.Lock()
+	defer manager.Unlock()
+
+	_, ok := manager.pendingEndpointEvents[id]
+	return ok
+}
+
+func (manager *Manager) addEndpoint(id types.NamespacedName) {
 	var epData *endpointMetadata
 	var err error
 	var identityLabels labels.Labels
@@ -242,20 +322,33 @@ func (manager *Manager) OnUpdateEndpoint(endpoint *k8sTypes.CiliumEndpoint) {
 	manager.Lock()
 	defer manager.Unlock()
 
+	endpoint, ok := manager.pendingEndpointEvents[id]
+	if !ok {
+		// the endpoint event has been already processed (for example we
+		// received an updated endpoint object or a delete event),
+		// nothing to do
+		return
+	}
+
 	logger := log.WithFields(logrus.Fields{
 		logfields.K8sEndpointName: endpoint.Name,
 		logfields.K8sNamespace:    endpoint.Namespace,
 	})
 
-	if len(endpoint.Networking.Addressing) == 0 {
+	if identityLabels, err = manager.getIdentityLabels(uint32(endpoint.Identity.ID)); err != nil {
 		logger.WithError(err).
-			Error("Failed to get valid endpoint IPs, skipping update to egress policy.")
+			Warning("Failed to get identity labels for endpoint")
 		return
 	}
 
-	if identityLabels, err = manager.getIdentityLabels(uint32(endpoint.Identity.ID)); err != nil {
+	// delete the endpoint from pendingEndpointEvent, from now on if we
+	// encounter a failure it cannot be retried
+	delete(manager.pendingEndpointEvents, id)
+
+	if len(endpoint.Networking.Addressing) == 0 {
 		logger.WithError(err).
-			Error("Failed to get identity labels for endpoint, skipping update to egress policy.")
+			Error("Failed to get valid endpoint IPs, skipping update to egress policy.")
+
 		return
 	}
 
@@ -265,9 +358,29 @@ func (manager *Manager) OnUpdateEndpoint(endpoint *k8sTypes.CiliumEndpoint) {
 		return
 	}
 
+	if _, ok := manager.epDataStore[epData.id]; ok {
+		logger.Debug("Updated CiliumEndpoint")
+	} else {
+		logger.Debug("Added CiliumEndpoint")
+	}
+
 	manager.epDataStore[epData.id] = epData
 
 	manager.reconcile(eventUpdateEndpoint)
+}
+
+// OnUpdateEndpoint is the event handler for endpoint additions and updates.
+func (manager *Manager) OnUpdateEndpoint(endpoint *k8sTypes.CiliumEndpoint) {
+	id := types.NamespacedName{
+		Name:      endpoint.GetName(),
+		Namespace: endpoint.GetNamespace(),
+	}
+
+	manager.Lock()
+	defer manager.Unlock()
+
+	manager.pendingEndpointEvents[id] = endpoint
+	manager.endpointEventsQueue.Add(id)
 }
 
 // OnDeleteEndpoint is the event handler for endpoint deletions.
@@ -275,12 +388,20 @@ func (manager *Manager) OnDeleteEndpoint(endpoint *k8sTypes.CiliumEndpoint) {
 	manager.Lock()
 	defer manager.Unlock()
 
+	logger := log.WithFields(logrus.Fields{
+		logfields.K8sEndpointName: endpoint.Name,
+		logfields.K8sNamespace:    endpoint.Namespace,
+	})
+
 	id := types.NamespacedName{
 		Name:      endpoint.GetName(),
 		Namespace: endpoint.GetNamespace(),
 	}
 
 	delete(manager.epDataStore, id)
+	delete(manager.pendingEndpointEvents, id)
+
+	logger.Debug("Deleted CiliumEndpoint")
 
 	manager.reconcile(eventDeleteEndpoint)
 }

--- a/pkg/egressgateway/manager_privileged_test.go
+++ b/pkg/egressgateway/manager_privileged_test.go
@@ -5,8 +5,10 @@ package egressgateway
 
 import (
 	"context"
+	"fmt"
 	"net"
 	"testing"
+	"time"
 
 	. "github.com/cilium/checkmate"
 	"k8s.io/apimachinery/pkg/types"
@@ -528,6 +530,18 @@ func parseIPRule(sourceIP, destCIDR, egressIP string, ifaceIndex int) parsedIPRu
 }
 
 func assertIPRules(c *C, rules []ipRule) {
+	var err error
+	for i := 0; i < 10; i++ {
+		if err = tryAssertIPRules(rules); err == nil {
+			break
+		}
+		time.Sleep(10 * time.Millisecond)
+	}
+
+	c.Assert(err, IsNil)
+}
+
+func tryAssertIPRules(rules []ipRule) error {
 	parsedRules := []parsedIPRule{}
 	for _, r := range rules {
 		parsedRules = append(parsedRules, parseIPRule(r.sourceIP, r.destCIDR, r.egressIP, r.ifaceIndex))
@@ -544,12 +558,15 @@ nextRule:
 			if rule.sourceIP.Equal(installedRule.Src.IP) && rule.destCIDR.String() == installedRule.Dst.String() &&
 				rule.ifaceIndex == installedRule.Table-linux_defaults.RouteTableEgressGatewayInterfacesOffset {
 
-				assertIPRoutes(c, rule.egressIP, rule.ifaceIndex)
+				if err := tryAssertIPRoutes(rule.egressIP, rule.ifaceIndex); err != nil {
+					return err
+				}
+
 				continue nextRule
 			}
 		}
 
-		c.Fatal("Missing IP rule")
+		return fmt.Errorf("Missing IP rule")
 	}
 
 nextInstalledRule:
@@ -561,11 +578,13 @@ nextInstalledRule:
 			}
 		}
 
-		c.Fatal("Untracked IP rule")
+		return fmt.Errorf("Untracked IP rule")
 	}
+
+	return nil
 }
 
-func assertIPRoutes(c *C, egressIP net.IPNet, ifaceIndex int) {
+func tryAssertIPRoutes(egressIP net.IPNet, ifaceIndex int) error {
 	eniGatewayIP := getFirstIPInHostRange(egressIP)
 	routingTableIdx := egressGatewayRoutingTableIdx(ifaceIndex)
 
@@ -577,7 +596,7 @@ func assertIPRoutes(c *C, egressIP net.IPNet, ifaceIndex int) {
 	}, netlink.RT_FILTER_OIF|netlink.RT_FILTER_DST|netlink.RT_FILTER_SCOPE|netlink.RT_FILTER_TABLE)
 
 	if err != nil || route == nil {
-		c.Fatal("Cannot find nexthop route to the VPC subnet:", err)
+		return fmt.Errorf("Cannot find nexthop route to the VPC subnet: %s", err)
 	}
 
 	route, err = netlink.RouteListFiltered(netlink.FAMILY_V4, &netlink.Route{
@@ -586,8 +605,10 @@ func assertIPRoutes(c *C, egressIP net.IPNet, ifaceIndex int) {
 	}, netlink.RT_FILTER_DST|netlink.RT_FILTER_TABLE|netlink.RT_FILTER_GW)
 
 	if err != nil || route == nil {
-		c.Fatal("Cannot find default route to the VPC:", err)
+		return fmt.Errorf("Cannot find default route to the VPC: %s", err)
 	}
+
+	return nil
 }
 
 func parseEgressRule(sourceIP, destCIDR, egressIP, gatewayIP string) parsedEgressRule {
@@ -620,6 +641,18 @@ func parseEgressRule(sourceIP, destCIDR, egressIP, gatewayIP string) parsedEgres
 }
 
 func assertEgressRules(c *C, policyMap egressmap.PolicyMap, rules []egressRule) {
+	var err error
+	for i := 0; i < 10; i++ {
+		if err = tryAssertEgressRules(policyMap, rules); err == nil {
+			break
+		}
+		time.Sleep(10 * time.Millisecond)
+	}
+
+	c.Assert(err, IsNil)
+}
+
+func tryAssertEgressRules(policyMap egressmap.PolicyMap, rules []egressRule) error {
 	parsedRules := []parsedEgressRule{}
 	for _, r := range rules {
 		parsedRules = append(parsedRules, parseEgressRule(r.sourceIP, r.destCIDR, r.egressIP, r.gatewayIP))
@@ -627,12 +660,20 @@ func assertEgressRules(c *C, policyMap egressmap.PolicyMap, rules []egressRule) 
 
 	for _, r := range parsedRules {
 		policyVal, err := policyMap.Lookup(r.sourceIP, r.destCIDR)
-		c.Assert(err, IsNil)
+		if err != nil {
+			return fmt.Errorf("cannot lookup policy entry: %w", err)
+		}
 
-		c.Assert(policyVal.GetEgressIP().Equal(r.egressIP), Equals, true)
-		c.Assert(policyVal.GetGatewayIP().Equal(r.gatewayIP), Equals, true)
+		if !policyVal.GetEgressIP().Equal(r.egressIP) {
+			return fmt.Errorf("mismatched egress IP")
+		}
+
+		if !policyVal.GetGatewayIP().Equal(r.gatewayIP) {
+			return fmt.Errorf("mismatched gateway IP")
+		}
 	}
 
+	untrackedRule := false
 	policyMap.IterateWithCallback(
 		func(key *egressmap.EgressPolicyKey4, val *egressmap.EgressPolicyVal4) {
 			for _, r := range parsedRules {
@@ -641,6 +682,13 @@ func assertEgressRules(c *C, policyMap egressmap.PolicyMap, rules []egressRule) 
 				}
 			}
 
-			c.Fatal("Untracked egress policy")
+			untrackedRule = true
+			return
 		})
+
+	if untrackedRule {
+		return fmt.Errorf("Untracked egress policy")
+	}
+
+	return nil
 }

